internal/kafka/internal/services/quota/internal/utils/kafka_billing_model_resolver.go

```diff
@@ -0,0 +1,118 @@
+package utils
+
+import (
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/api/dbapi"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/internal/kafka/internal/config"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/errors"
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/shared/utils/arrays"
+)
+
+/********************************************************************************************************************
+ * This resolver is used when the user specifies a `DesiredBillingModel` in the request.
+ * If the specified `DesiredBillingModel` is a `marketplace`, attempts are done to detect the marketplace type (if
+ * not specified)
+ ********************************************************************************************************************/
+
+var _ BillingModelResolver = &kafkaBillingModelResolver{}
+
+type kafkaBillingModelResolver struct {
```

Comment:
Is there a difference in behavior here?

Before this PR checking whether there is quota available is performed whereas in kafka billing model resolver it is unconditionally returned without checking it

Reply:
Can you elaborate more? Before this PR, the code to resolve the billing model was in 3 main `if` clauses.
I split that code into three objects (resolvers), but the main logic should be the same.
What is the check that is missing now?

Reply:
The check that might be missing is whether there is quota availale

Reply:
The kafka billing model resolver is used when the `user` explicitly asks for a spcific kafka billing model (DesiredBillingModel is not nil).

If the requested billing model supports only one ams billing model, there is no need to make any detection, since that is exactly what the user asked for. If there is no quota for that, it will fail to reserve the quota.

If the requested billing model supports more than one ams billing model, then the detection is triggered, and the first ams billing model for whom the user has quota for is returned.

There is no change in behaviour: if the user has not enough quota, it will return a no quota error in both cases.

Reply:
> If the requested billing model supports only one ams billing model, there is no need to make any detection, since that is exactly what the user asked for. If there is no quota for that, it will fail to reserve the quota.

That is the change in behavior I referred to.

Maybe "change in behavior" was not specific enough: The user would experience the same behavior although a different error would be returned.

Before, the quota check was explicitly performed before reserving quota and now it is delegated to the AMS API.

In other resolvers this check continues to be explicitly performed before sending the request to the AMS API